### PR TITLE
Add a new configuration for L018 to ignore indentations

### LINKED ISSUE #3498: Add configuration to ignore indentation in L018
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

In case CTE is inside template block, the L018 and L003 contradict each other.

### Use case

```
with
{% if true %}
  cte as (
      select 1
  )
{% else %}
  cte as (
      select 2
  )
{% endif %}
select * from cte
```

I won't expect any lint errors here, but L018 suggests to de-indent closing brackets, and L003 suggests to indent them back.

### Dialect

No specific dialect

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by zidder**
### Brief summary of the change made

fixes #3498 

Add a new configuration variable for rule L018 to ignore indentations. This will be helpful in case the CTE is in a template block. In that case, currently L018 fails suggesting to move closing bracket, which contradicts to L003. The new solution basically suggests to move closing bracket to a new line, but to not look at the indent, L003 will take care of indentations.

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by tunetheweb** [comment]
I\m a bit confused with this change. What's the difference between ignoring indents and just turning off the rule?

If `ignore_indents` is set to True, won't this basically mean the rule never runs?

**Comment by zidder** [comment]
> I\m a bit confused with this change. What's the difference between ignoring indents and just turning off the rule?
> 
> If `ignore_indents` is set to True, won't this basically mean the rule never runs?

Actually no, see the `test_pass_with_clause_closing_misaligned_indentation_ignoring_indent` test in yaml tests. It fails with both `ignore_indent` and without. The idea is to move the closing bracket to the new line, but not check indents and leave it to L003. If you check the L018.py, there are two cases where the rule fails, and one of them is turned off in case `ignore_indent=True`.

**Comment by zidder** [comment]
I see the issue now. I was thinking about linting only, and lints will pass only for the nice SQL:
```
    with
    {% if true %}
      cte as (
          select 2
      )
    {% else %}
      cte as (
          select 2
      )
    {% endif %}
    select * from cte
```

But the fix doesn't work properly. I'll make this work if I can.
Thanks for reviewing.

**Comment by tunetheweb** [comment]
So I guess you're going to try to indent that properly? In that case maybe the config name should be changed from `ignore_indents` to `allow_indents`? In which case the first case should check the correct number of whitespaces have been given as well, rather than just any whitespaces?

I'm still not convinced it's worth it, rather than just disabling the rule. But if you're happy to work on it, and think you can get something clean and reasonably easy to code and maintain then happy to review once you do.

**Comment by zidder** [comment]
@tunetheweb I found the reason the fix didn't work properly, and I'm not happy with it :( It tried to apply fix to the initial file during the 3rd iteration of the loop.

**Comment by zidder** [comment]
@barrywhart 
Do you mean changing logic of the rule? Are you sure this won't break somebody else's code (SQL) ?
I'll be satisfied with that solution if there won't be any negative effects.
I'm still not very familiar with the code, but I'll try to change it properly.

**Comment by barrywhart** [comment]
That's what I mean. I haven't thought about this option carefully, but as @tunetheweb says, if we can find a solution we're happy with, that sounds good.

**Comment by zidder** [comment]
@barrywhart , Changed L018 so that it won't take responsibility to fix indentations.

Tested with the following SQLs when L003 is enabled.

First SQL:
```
WITH
{% if true %}
    cte AS (
        SELECT 1)
{% else %}
    cte AS (
        SELECT 2)
{% endif %}
SELECT * FROM cte
```

First Fixed SQL:
```
WITH
{% if true %}
    cte AS (
        SELECT 1
    )
{% else %}
    cte AS (
        SELECT 2)
{% endif %}
SELECT * FROM cte
```

Second SQL:
```
WITH
{% if true %}
    cte AS (
        SELECT 2
)
{% endif %}

SELECT * FROM cte
```

Second Fixed SQL:
```
WITH
{% if true %}
    cte AS (
        SELECT 2
    )
{% endif %}

SELECT * FROM cte
```

Third SQL:
```
WITH cte AS (
    SELECT 1
    ) SELECT * FROM cte
```

Third Fixed SQL:
```
WITH cte AS (
    SELECT 1
) SELECT * FROM cte
```

Forth SQL:
```
    WITH cte AS (
        SELECT 1
)

SELECT * FROM cte
```

Forth Fixed SQL:
```
WITH cte AS (
    SELECT 1
)

SELECT * FROM cte
```

As you can see, those SQLs were fixed with L003 and L018 combined, and L018 doesn't contradict L003 anymore. Test cases don't show fixes by L003, but you can see, that L018 adds new lines only.

**Comment by tunetheweb** [comment]
Note you can do combination tests (e.g. L018 and L003) with a python test.
Have a look here at this example: https://github.com/sqlfluff/sqlfluff/blob/main/test/rules/std_L003_L065_combo_test.py

**Comment by barrywhart** [comment]
You can also do combination tests without writing Python. See `test/rules/std_fix_auto_test.py`. It picks up test cases from the `test/fixtures/linter/autofix` directory.

**Comment by barrywhart** [comment]
Actually, one of those `autofix` tests is failing: `test/rules/std_fix_auto_test.py::test__std_fix_auto[ansi-007_with_clause]`

File: test/fixtures/linter/autofix/ansi/007_with_clause/after.sql [hunk 1]
```diff
@@ -1,5 +1,5 @@
 -- Dealing with complicated indents before with clauses.
     	WITH cte as (
             select a from tbla
-        )
+)
         select a from cte
```
Comment (tunetheweb) [inline test/fixtures/linter/autofix/ansi/007_with_clause/after.sql:5]:
```suggestion
        select a from cte

```

File: test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+WITH
+{% if true %}
+    cte AS (
+        SELECT 1
+    )
+{% else %}
+    cte AS (
+        SELECT 2)
```
Comment (tunetheweb) [inline test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql:8]: Should this not be fixed too?

File: test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+WITH cte AS (
+    SELECT 1
+) SELECT * FROM cte
```
Comment (tunetheweb) [inline test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql:3]: Should the rule be changed to be on a new line of it's own?

To fix this to:

```sql
WITH cte AS (
    SELECT 1
)

SELECT * FROM cte
```

File: test/fixtures/rules/std_rule_cases/L018.yml [hunk 4]
```diff
@@ -11,28 +11,19 @@ test_pass_with_clause_closing_oneline:
...
   pass_str: with cte as (select 1) select * from cte
 
-test_fail_with_clause_closing_misaligned_indentation:
+test_pass_with_clause_closing_misaligned_indentation:
   # Fix with statement indentation
-  fail_str: |
+  pass_str: |
     with cte as (
         select 1
         ) select * from cte
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L018.yml:19]: To me this is a regression. This should fail IMHO and be fixed.

Instead of aligning with the `WITH` should we align with wherever the CTE opening bracket line starts? (which in this case is at the start of the `WITH`, but in your templated example might not be.).

File: test/fixtures/rules/std_rule_cases/L018.yml [hunk 5]
```diff
@@ -11,28 +11,19 @@ test_pass_with_clause_closing_oneline:
...
-    ) select * from cte
-
-test_fail_with_clause_closing_misaligned_negative_indentation:
+test_pass_with_clause_closing_misaligned_negative_indentation:
   # Fix with statement that has negative indentation
-  fail_str: |2
+  pass_str: |2
         with cte as (
         select 1
     ) select * from cte
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L018.yml:26]: Again I see this as a regression due to this change - this should be fixed.

File: test/fixtures/rules/std_rule_cases/L018.yml [hunk 6]
```diff
@@ -59,3 +50,42 @@ test_pass_cte_with_column_list:
...
+
+  fix_str: |
+    with
+    {% if true %}
+      cte as (
+          select 1
+    )
+    {% else %}
+      cte as (
+          select 2)
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L018.yml:89]: Again I feel this line should be fixed.

File: src/sqlfluff/rules/L018.py [hunk 7]
```diff
@@ -22,25 +21,23 @@
 @document_fix_compatible
 @document_configuration
 class Rule_L018(BaseRule):
-    """``WITH`` clause closing bracket should be aligned with ``WITH`` keyword.
+    """``WITH`` clause closing bracket should be on a new line.
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L018.py:24]: As mentioned elsewhere I think the rule description should be changed to "Common Table Expressions should have the closing bracket aligned to the opening bracket line". Or something like that.

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: src/sqlfluff/rules/L018.py (hunk 7 above)
Comment (tunetheweb) [reply src/sqlfluff/rules/L018.py:24]: Ah I missed that. In that case a lot of my review may be rubbish :-)

File: test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql (hunk 2 above)
Comment (barrywhart) [reply test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql:8]: No, because this part of the template is not rendered (it's in the `else` block of an `if true`.) If this `if` statement were inside a loop that at times evaluated to both `true` and `false`, then SQLFluff would "see" this code and fix it. It's an odd side effect of the two phase linting (templating, then parsing).

File: test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql (hunk 3 above)
Comment (barrywhart) [reply test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql:3]: Interesting. Is there a rule which does that? It seems out of scope for this PR. 🤔

File: test/fixtures/rules/std_rule_cases/L018.yml (hunk 4 above)
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L018.yml:19]: IIUC, this PR is updating L018 so it doesn't concern itself with indentation -- it leaves that job to L003. Now L018 is **only** responsible for making sure the closing bracket of a CTE definition is on a separate line.

File: test/fixtures/rules/std_rule_cases/L018.yml (hunk 5 above)
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L018.yml:26]: Seems like it's okay because the `)` is on a line by itself, and that's all L018 will be responsible for after this PR.

File: test/fixtures/linter/autofix/ansi/007_with_clause/after.sql [hunk 8]
```diff
@@ -1,5 +1,5 @@
 -- Dealing with complicated indents before with clauses.
     	WITH cte as (
             select a from tbla
-        )
+)
```
Comment (tunetheweb) [inline test/fixtures/linter/autofix/ansi/007_with_clause/after.sql:4]: I think this test case is a bit pointless. It's only one rule. It shouldn't really be in this folder and should be in the rule YAML file.

However, now we are making L003 do the indentation I think we could leave it here, if we added L003 to the test config and then make sure that the closing bracket is aligned. In which case we can revert this change.

File: test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql (hunk 2 above)
Comment (tunetheweb) [reply test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql:8]: OK makes sense, but not obvious. Can we add a comment to that effect?

File: test/fixtures/rules/std_rule_cases/L018.yml (hunk 4 above)
Comment (tunetheweb) [reply test/fixtures/rules/std_rule_cases/L018.yml:19]: Ignore. L003 will handle this. Resolving.

File: test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql (hunk 2 above)
Comment (barrywhart) [reply test/fixtures/linter/autofix/ansi/027_l003_l018_templated_block_without_newline_cte/after.sql:8]: Could we change this test to include a loop and an if? Something like:
```
WITH
{% for i in [0, 1] %}
    {% if i == 0 %}
        cte0 AS (
            SELECT 1
        )
    {% else %}
        cte1 AS (
            SELECT 2)
    {% endif %}
{% endfor %}
SELECT * FROM cte1
UNION ALL
SELECT * FROM cte2
```

File: test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql (hunk 3 above)
Comment (tunetheweb) [reply test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql:3]: Good point. L022 handles that. Maybe should be added to this test config?

Comment (barrywhart) [reply test/fixtures/linter/autofix/ansi/027_l003_l018_wrong_indent_cte/after.sql:3]: Sounds good to me. I see these tests as narrowly defined real-world tests, i.e. choose the minimum set of rules to do a nice job on the query.

File: test/fixtures/linter/autofix/ansi/007_with_clause/after.sql (hunk 1 above)
Comment (zidder) [reply test/fixtures/linter/autofix/ansi/007_with_clause/after.sql]: Added the `newline` in both `before.yml` and `after.yml`

File: test/fixtures/linter/autofix/ansi/007_with_clause/after.sql (hunk 8 above)
Comment (zidder) [reply test/fixtures/linter/autofix/ansi/007_with_clause/after.sql]: Added the L003 and changed the `after.yml`

**Comment by tunetheweb** [comment]
Good work @zidder !

**Comment by zidder** [comment]
Thanks for being patient and reviewing my pull request :)
